File: Tree/Binary Tree/Binary_Search_Tree/BST.java
import java.util.*;

class BST{
    class Node{
        int data;
        Node left;
        Node right;

        Node(int data){
            this.data = data;
            this.left = null;
            this.right = null;
        }
    }

    Node root;

    private Node BuildTree(ArrayList<Integer> arr, int s, int e){
        if(s > e) return null;
        int m = (s+e)/2;
        Node root = new Node(arr.get(m));
        root.left = BuildTree(arr, s, m-1);
        root.right = BuildTree(arr, m+1, e);
        return root;
    }

    private void insert(Node root, int data){
        if(root.data > data){
            if(root.left != null){
                insert(root.left, data);
            }else{
                root.left = new Node(data);
            }
        }else{
            if(root.right != null){
                insert(root.right, data);
            }else{
                root.right = new Node(data);
            }
        }
    }

    BST(){
        this.root = null;
    }

    BST(ArrayList<Integer> arr){
        this.root = BuildTree(arr, 0, arr.size()-1);
    }

    void insert(int data){
        if(root == null){
            root = new Node(data);
            return;
        }

        insert(root, data);
    }


    void printTree(){
        if(root == null){
            return;
        }
        Queue<Node> q = new LinkedList<>();
        q.add(root);
        while(!q.isEmpty()){
            int n = q.size();
            while(n-- > 0){
                Node curr = q.poll();
                System.out.print(curr.data + " ");
                if(curr.left != null) q.add(curr.left);
                if(curr.right != null) q.add(curr.right);
            }
            System.out.println();
        }
    }

    public static void main(String []args){
        BST b = new BST();

        ArrayList<Integer> arr = new ArrayList<>(Arrays.asList(1,2,3,4,5,6,7));
        BST b1 = new BST(arr);


        b.insert(5);
        b.insert(2);
        b.insert(9);
        b.insert(1);
        b.insert(4);
        b.insert(7);
        b.insert(14);
        b.insert(3);
        b.insert(6);
        b.insert(8);
        b.insert(11);

        b1.insert(8);

        b1.printTree();

    }
}

File: Tree/Binary Tree/Binary_Search_Tree/BST.cpp
#include<bits/stdc++.h>
using namespace std;


class BST{
    class Node{
      private:
        int val, height;
        Node *left, *right;

      public:
        Node(int val){
            this->val = val;
            this->height = 1;
            this->left = this->right = nullptr;
        }

        int getVal(){return this->val;}

        void setVal(int val){this->val = val;}

        int getHeight(){return this->height;}

        void setHeight(int height){this->height = height;}

        void setLeft(Node *left){this->left = left;}

        void setRight(Node *right){this->right = right;}
        
        Node* getLeft(){return this->left;}

        Node* getRight(){return this->right;}
    };

    Node *root;

    Node* createBSTWithSortedArray(vector<int> arr, int start, int end){
        if(start > end) return NULL;

        int mid = (start+end)/2;
        Node *root = new Node(arr[mid]);

        root->setLeft(createBSTWithSortedArray(arr,start,mid-1));
        root->setRight(createBSTWithSortedArray(arr,mid+1,end));

        return root;
    }

    void insertValue(Node *root, int val){
        if(val < root->getVal()){
            if(root->getLeft() == nullptr){
                root->setLeft(new Node(val));
                return;
            }
            insertValue(root->getLeft(),val);
        }else{
            if(root->getRight() == nullptr){
                root->setRight(new Node(val));
                return;
            }
            insertValue(root->getRight(),val);
        }
    }

    Node* deleteNode(Node* root, int val) {
        if (root == nullptr) 
            return root;

        if (val < root->getVal()) {
            root->setLeft(deleteNode(root->getLeft(), val));
        } 
        else if (val > root->getVal()) {
            root->setRight(deleteNode(root->getRight(), val));
        } 
        else {
            // Case 1: no child
            if (root->getLeft() == nullptr && root->getRight() == nullptr) {
                delete root;
                return nullptr;
            }

            // Case 2: one child
            else if (root->getLeft() == nullptr) {
                Node* temp = root->getRight();
                delete root;
                return temp;
            }

            else if (root->getRight() == nullptr) {
                Node* temp = root->getLeft();
                delete root;
                return temp;
            }

            // Case 3: two children
            else {
                Node* temp = root->getRight();
                while (temp->getLeft() != nullptr)
                    temp = temp->getLeft();

                root->setVal(temp->getVal());
                root->setRight(deleteNode(root->getRight(), temp->getVal()));
            }
        }
        return root;
    }


    void print2DTree(Node* root, int space = 0, int indent = 8) {
        if (root == nullptr) return;

        // Increase distance between levels
        space += indent;

        // Print right child first
        print2DTree(root->getRight(), space);

        // Print current node after spaces
        cout << string(space - indent, ' ') << root->getVal() << "\n";

        // Print left child
        print2DTree(root->getLeft(), space);
    }

    public:
    BST(){
        this->root = nullptr;
    }

    BST(vector<int> arr){
        this->root = createBSTWithSortedArray(arr,0,arr.size()-1);
    }


    void insertValue(int val){
        if(root == nullptr){
            root = new Node(val);
            return;
        }
        insertValue(root,val);
    }

    void deleteNode(int val){
        root = deleteNode(root, val);
    }

    void print2DTree() {
        print2DTree(root);
    }

    void printTree() {
        if (root == nullptr) {
            cout << "null\n";
            return;
        }

        queue<Node*> q;
        q.push(root);

        bool hasNonNull = true;  // track if further levels contain non-null nodes

        while (hasNonNull) {
            int n = q.size();
            hasNonNull = false;

            for (int i = 0; i < n; i++) {
                Node* temp = q.front();
                q.pop();

                if (temp == nullptr) {
                    cout << "null ";
                    q.push(nullptr);
                    q.push(nullptr);
                } else {
                    cout << temp->getVal() << " ";
                    q.push(temp->getLeft());
                    q.push(temp->getRight());

                    if (temp->getLeft() != nullptr || temp->getRight() != nullptr)
                        hasNonNull = true;
                }
            }
            cout << endl;
        }
    }

};


int main(){

    BST b;
    b.insertValue(5);
    b.insertValue(2);
    b.insertValue(9);
    b.insertValue(1);
    b.insertValue(3);
    b.insertValue(7);
    b.insertValue(10);
    b.insertValue(4);
    b.insertValue(6);
    b.insertValue(8);

    b.printTree();
    b.print2DTree();

    cout << "\n\n After deleting a Number 9:\n";
    b.deleteNode(9);
    b.printTree();
    b.print2DTree();

    
    return 0;
}